src/lake/Lake/Load/Elab.lean

```diff
@@ -76,51 +77,113 @@ def elabConfigFile (pkgDir : FilePath) (lakeOpts : NameMap String)
 We call it here via `@[extern]` with a mock implementation.
 -/
 @[extern "lean_environment_add"]
-private def add (env : Environment) (_ : ConstantInfo) : Environment := env
+private opaque addToEnv (env : Environment) (_ : ConstantInfo) : Environment
 
 /--
-Import the OLean for the configuration file if `reconfigure` is not set
-and an up-to-date one exists (i.e., one newer than the configuration and the
-toolchain). Otherwise, elaborate the configuration and save it to the OLean.
+Imports a configuration `.olean` (e.g., `lakefile.olean`).
+Auxiliary definition for `importConfigFile`.
 -/
-def importConfigFile (wsDir pkgDir lakeDir : FilePath) (lakeOpts : NameMap String)
+def importConfigFileCore (olean : FilePath) (leanOpts : Options) : IO Environment := do
+  /-
+  We could import the olean together with its imports using one
+  `importModules` call, but that would mean we pay for a full
+  `finalizeImports` each time, which is linear in the number of imported
+  constants and extension entries (in fact, it is paid two more times: when
+  marking the `Environment` object as multi-threaded, and when releasing
+  it). As most lakefiles use the same set of imports, we instead construct
+  an `Environment` for it only once, and then apply the lakefile contents
+  on top of it like the elaborator would. Thus the non-shared, part of the
+  `Environment` is very small.
+  -/
+  let (mod, _) ← readModuleData olean
+  let mut env ← importModulesUsingCache mod.imports leanOpts 1024
+  -- Apply constants (does not go through the kernel, so order is irrelevant)
+  env := mod.constants.foldl addToEnv env
+  /-
+  Apply extension entries (`PersistentEnvExtension.addEntryFn` is pure and
+  does not have access to the whole environment, so no dependency worries
+  here either
+  -/
+  let extDescrs ← persistentEnvExtensionsRef.get
+  let extNameIdx ← mkExtNameMap 0
+  for (extName, ents) in mod.entries do
+    if let some entryIdx := extNameIdx.find? extName then
+      for ent in ents do
+        env := extDescrs[entryIdx]!.addEntry env ent
+  return env
+
+instance : ToJson Hash := ⟨(toJson ·.val)⟩
+instance : FromJson Hash := ⟨((⟨·⟩) <$> fromJson? ·)⟩
+
+structure ConfigTrace where
+  platform : String
+  leanHash : String
+  configHash : Hash
+  options : NameMap String
+  deriving ToJson, FromJson
+
+inductive ConfigLock
+| olean (h : IO.FS.Handle)
+| lean (h : IO.FS.Handle) (lakeOpts : NameMap String)
+
+/--
+Import the `.olean` for the configuration file if `reconfigure` is not set and
+an up-to-date one exists (i.e., one with matching configuration and on the same
+toolchain). Otherwise, elaborate the configuration and save it to the `.olean`.
```

Comment:
```suggestion
Imports the `.olean` for the configuration file if `reconfigure` is not set and
an up-to-date one exists (i.e., one with matching configuration and on the same
toolchain). Otherwise, elaborates the configuration and saves it to the `.olean`.
```

Reply:
I did not do this one for now, because the rest of the file is currently in the imperative mood. I think this change would be best done in a full sweep refactor.